### DISCUSSION TITLE
New unified alerting reduce options are not enough

### DISCUSSION THREAD

**Discussion by Owersun** [grafana/grafana#42771] [Alerting]
The new unified alerting rules that needs to be used to have labels on alerts do not have enough flexibility with current reduce options.
There is only Min/Max/Mean/Sum/Count which reduces time span to a value and do not allow to evaluate the change over the time period.
Example:
We have queue system with many separate queues. Messages in them are processed faster than they go in, so most of them are 0 all the time. Sometimes a huge chunk of events can arrive which will make a queue go sky high with numbers. That's not an issue by itself, so I'm not interested in queue being > 0. As I mentioned the messages are processed faster than they arrive, so normal situation is that a queue numbers goes down over time. An abnormal situation that I want to monitor is that queue number does not go down in any short span, like 3-5 minutes. Messages are processed fast, any situation that queue has increasing number of messages or _same_ amount of messages for 3-5 minutes will indicate the queue is stuck.
With current reduce options it's very hard to do that properly, Min/Max/Mean/Sum/Count they all don't give me overview around the "change" happening over time span. Evaluating Min to Max does not give any advantage either, cause it's not clear does the queue go up or down, Min and Max don't have time relativity to each other. Making the query as "rate" or "increase" also doesn't work. For some reason it's never negative. It either indicates messages amount go up or flat, and "flat" can be as "flat" at N, which is a problem (queue is stuck), or "flat" at 0, which is not a problem (queue is empty).
So far I was able to work it around by making the alarm evaluate two queries: one "now to now - 2m" and other "now - 2m to now - 4m", reducing them both to avg number, and comparing clearly how many messages were there 2 to 4 minutes ago, and now to 2 minutes ago. That make it possible to say does the number go down or not. But this really feels like ugly workaround, taking two time spans to reduce them to a single number, and compare to each other, instead of having a number for start of the span and end of the span?.. Sounds like a dirty hack.
It would be really nice if something like "Diff" or "First"/"Last" as Reduce option is added.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
